### DISCUSSION TITLE
QualCoder 3.5

### DISCUSSION THREAD

**Discussion by ccbogel** [ccbogel/QualCoder#876] [Announcements]
## Early release of 3.5

This an early release of 3.5 to have all current updates and fixes in place, as there is potentially a major change coming in the next version, with introduction of artificial intelligence and GPT4 identification of coded text. (Still in development thanks to Kai Droge).

A binary (double-click to run) for Windows is available at the end of the release page for:
QualCoder-3.5.exe md5 checksum: 3faa37a813bf3944b2409c424077ee19

On first use of the exe, Windows will ask you to allow to run QualCoder. This is because it is from an unknown publisher. It costs a lot of money to get a trusted publisher certificate - so that will not be possible for the foreseeable future.

A binary (double click to run) is available for Ubuntu 22.04. The md5 checksum is: 

For install from source code, download the zip file below and use the instructions on the main GitHub QualCoder page, or the [wiki](https://github.com/ccbogel/QualCoder/wiki/01-Introduction) to install on your operating system.

## Report codes
Removed [VIEW] in the header of all the code segments in results. It has been replaced with character positions for coded text segments. e.g. [150-188]

Added context menu option -  Change code to another code.

Replaced cogs button with Play button triangle. Repositioned and made larger button. Replace word 'search' with 'Run report'.

## NEW Report - Code text exact matches

Added a new report that reports on exacted text matches in each file, where multi0le codes have been applied to the exact same text. This report is perhaps particularly useful for those people who use codes as variables to perform analyses.

## Code text
Export html now includes tooltips when hovering over coded sections. Some character set display issues for some characters, e.g. Euro symbol.
Export text is a new option. this exports a text file with coded segment marked with text inn parentheses like this:
Regular text {{code name{{the tagged text portion}}code name}} regular text.

Stating character position: click in the displayed text and press ! to show the numeric position in the text. Press $ to shift all coded segments after the clicked position by up to to +/- 500 characters.

## Charts - Word cloud
The word cloud has been changed with 'home grown' code. That is, I made my own word cloud code in python. This was mainly to prevent errors with packaging the PyPi Wordcloud into pyinstaller.

## Error fixes
* Zero division error in compare coders. Occurs under a specific circumstance - a code has been created but never assigned to data. Fixed.

* Creating a Case Attribute from the Cases screen. An error occurs. A work around is to create the Attribute from the Manage Attributes screen. Fixed.

* Wordcloud function does not use coded text within selected category. Fixed.

<hr /><em>This discussion was created from the release <a href='https://github.com/ccbogel/QualCoder/releases/tag/3.5'>QualCoder 3.5</a>.</em>

**Comment by lkrs200** [comment]
Thank you for coding this alternative to paid programs!

I have a problem importing html files, it doesnt work at all. maybe i'm doing sth wrong idk...
When importing file in the manage tab it says:

Traceback:
File "qualcommander\manage_files.py", line 1356, in import_files
File "qualcoder\manage_files.py", line 1591, in load_file_text
File "encodings\cp1252.py", line 23, in decode

UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 707: character maps to <undefined>

**Reply by ccbogel** [reply]
Hi. Thanks for reporting that error. That means that some characters in the html cannot be read when the file is opened and it is crashing instead of working around the problem.
I have hopefully fixed the latest 3.6 code to deal with this. 
If you are using QC 3.5 I would suggest Copy the html file (open it in a web browser and select all) and Paste into notepad first, and save that notepad file. Then import the notepad file. Hope this helps.
If it still does not work - you could then try the same approach but instead of copy and paste into Notepad - you could try copy and paste into Word. But I think Notepad would be the better first step to try.

**Comment by Frangbg65** [comment]
Hi, I am trying to instal QualCoder on a M2 Mac, I am not particularly confident with using the terminal, I have followed all the steps indicated in the instructions, but when I try to run it, the program doesn't open, I always get this: /Users/franciscoferrer/anaconda3/lib/python3.11/site-packages/pydub/utils.py:170: RuntimeWarning: Couldn't find ffmpeg or avconv - defaulting to ffmpeg, but may not work
  warn("Couldn't find ffmpeg or avconv - defaulting to ffmpeg, but may not work", RuntimeWarning)
qt.qpa.plugin: Could not find the Qt platform plugin "cocoa" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Can you help?

Thanks

**Comment by ccbogel** [comment]
Hello Franciso,

I do not have a mac computer, so I cannot give heaps of help.

You can ignore the warning about ffmpeg. QualCoder will run without it.
It looks like you have used the terminal well.

The cocoa plugin is the problem.
I see several solutions, from google,  which may or may not work. As I am not familiar with macOS, I am not much help.

c++ - Qt could not find the platform plugin cocoa - Stack Overflow<https://stackoverflow.com/questions/54297627/qt-could-not-find-the-platform-plugin-cocoa>
The solution here is to run this line in the terminal, it is an older answer from 2019, it look simple to do, so its worth a try:


pip3 install opencv-python-headless

 python 3.x - qt.qpa.plugin: Could not find the Qt platform plugin "cocoa" in "" - Stack Overflow<https://stackoverflow.com/questions/76898551/qt-qpa-plugin-could-not-find-the-qt-platform-plugin-cocoa-in>
This solution requires you to find and export a QT_PLUNIN_PATH
It looks more complicated to try.

Do you know any mac users who are programmers?

with regards Colin

________________________________

**Comment by Frangbg65** [comment]
Hi Colin,

Much appreciated  for getting back to me. I have actually tried your first solution, it didn’t work. I am to a programmer, so the second solution looks too complicated for me. I’ll keep looking. I just wanted to use QualCoder on my thesis interviews, but I can’t spend so much time just trying to run the application. 

Thank you anyway.

Regards


Francisco

**Comment by kaixxx** [comment]
Hi Francisco

You can try my version of QualCoder. A friend of mine created an installer for macOS which may solve your problem (no promises, but worth a try): https://github.com/kaixxx/QualCoder/tree/ai_integration 
This is an experimental version with some additional AI-based funtionality. But you can just ignore the AI-stuff and use it as normal QualCoder.

**Comment by Frangbg65** [comment]
Hi Kaixxx,

Thank you,

I have tried, but it still asking for stage that I Donn’t know what it is as I’m not a developer. Thank you anyway. I think I’ll have to give QualCoder a miss.

Regards

Francisco

**Reply by kaixxx** [reply]
Have you used the installer "QualCoderAI_MAC_arm64.dmg"?

What do you mean by "asking for stage"? 
I understand that this might be too much hassle for you current project. But I would like to understand the problem so we can try to solve it for the future.

**Comment by KiwiAllan** [comment]
HI. I'm trying to install Qualcoder 3.5 into my Intel-based Macbook Pro, with Python 3.13. 

All OK, until I get to 'python3 /applications/QualCoder-3.5/qualcoder/__main__.py', and that ends up with the error code seen in the attached file. Are there any suggestions about how to resolve this?

Gratefully for any way forward.

Allan

![Screenshot 2024-11-05 at 12 01 43 PM](https://github.com/user-attachments/assets/da52f4ea-744f-420d-853b-9a3fa638bef6)

**Reply by ccbogel** [reply]
Hi. There seems to be a problem with the pydub module. My recommendation is to download the latest code zip (Green Code button on the main page). Then install again as you did before. but there will be no need to install pydub nor SpeechRecognition.
I have just now removed pydub and SpeechRecognition and the associated functionality from QualCoder. It is not needed anyway, as many other online and software can convert speech to text.

**Comment by KiwiAllan** [comment]
This is prompt and helpful. But I can't find 'a green code button on the main page'. 

Is the appropriate page: https://github.com/ccbogel/QualCoder/releases/tag/3.5? (which has the source file downoad in the Assets section). Or another page? Thanks for clarifying.

**Reply by ccbogel** [reply]
The main page is here: and the URL is shown also:
![image](https://github.com/user-attachments/assets/8914486c-82f4-419e-9782-da9a460e12a4)

**Reply by KiwiAllan** [reply]
Superb, thanks.

**Reply by KiwiAllan** [reply]
To run  3.6., what modifications do I need to make to this command (used for 3.5): python3 /applications/QualCoder-3.5/qualcoder/__main__.py

Gratefully.

**Reply by KiwiAllan** [reply]
... as simply changing '3.5' to '3.6' isn't recognised ('No such file or directory'). )-:

**Reply by ccbogel** [reply]
The downloaded folder via the Green button from GitHub would be called:
QualCoder-master
so the command should be:
python3 /applications/QualCoder-master/\_\_main\_\_.py

OR potentially from the same QualCoder-master folder:

python3 -m qualcoder

**Reply by KiwiAllan** [reply]
Unfortunately neither of these options work. The first gives me the 'No such file or directory' message; and the second the error message in the attached screenshot.
![Screenshot 2024-11-05 at 4 46 57 PM](https://github.com/user-attachments/assets/2b9d61d9-f9fa-42df-880b-f30fb49438f7)
.

**Comment by KiwiAllan** [comment]
For your interest and further suggestions... when I have run 'pip install -r requirements.txt' , it goes through well, except near the end where I get two error messages, as in the attached files.
![Screenshot 2024-11-05 at 1 54 35 PM](https://github.com/user-attachments/assets/9261791c-2145-4860-8ef5-de3bc219b0f1)
![Screenshot 2024-11-05 at 1 54 45 PM](https://github.com/user-attachments/assets/bbd50f59-fc04-4027-8507-363655b5847e)

**Reply by ccbogel** [reply]
Hmm. I wonder if it will run - providing you do not use the AI components (you need a key for AI chat GPT), and to activate AI in the Settings anyway. I provided the updated command above in response the an earlier question. I have asked Dr Droege about this conflict issue as he is developing the AI components of the software.
Try this first.
The other solution is that I edit the 3.5 code and send the zip folder of that for you to use. The edit will remove the use of pydub, the issue in the original problem that you posted.

**Reply by KiwiAllan** [reply]
I'm SO grateful you are investing time and expertise on this.

Actually, "The other solution is that I edit the 3.5 code and send the zip folder of that for you to use. The edit will remove the use of pydub, the issue in the original problem that you posted" will likely be the better option, because my colleague (on a PC) is using 3.5., and I wonder whether we might end up with compatibility issues if he;s on 3.5. and I'm on 3.6. Possible?

**Reply by kaixxx** [reply]
I will take a look at the issue with installing sentence-transformers.

**Reply by kaixxx** [reply]
Ok, I have tested the current requirements.txt on my PC in a new virtual environment and it installs fine. It downloads and installs torch 2.5.1 as a dependency. I am wondering if you have an old version of torch already installed on your system? 

@ccbogel EDIT: Never mind, I had not updated my local code properly, it is working fine now! 
[Original: There is another small issue now: The transcription function is still in the code of QualCoder 3.6, so running 3.6 in my newly created environment complains that pydub is missing. I agree that it is a good idea to remove the transcription feature since there are better options out there - including my own software noScribe ;) Would you do that for 3.6 too?]

**Reply by kaixxx** [reply]
And btw: I'm the "Dr Droege" Colin is referring to above. You can just call me Kai, Colin ;)

**Reply by KiwiAllan** [reply]
Thanks Kai. If you are asking me whether I have 'an old version of torch already installed on your system', I can't find that app in a Spotlight search on my Mac.

**Reply by kaixxx** [reply]
I think you won't find torch with the spotlight search. 
Instead, run `pip list` in a command window. You will get a long list of python module installed on your system and the corresponding version number. Look for "torch" and what version it is.

**Reply by kaixxx** [reply]
@ccbogel: Regarding the "small issue" with the transcription function I mentioned above - it is solved now. I had not updated my local code properly, sorry.

**Reply by ccbogel** [reply]
@KiwiAllan 
OK Here it is - this should work fine. It is the zipped qualcoder 3.5 (I added the suffix mod to show it has been modified). This removes the speech to text option, and in doing so, removes the need for the error-prone pydub module.
Regards 3.6 versus 3.5 comment. The main database (containing your coding and plain text representations of text files) is the same between the 2 versions - so interoperable.  And some errors/glitches have been fixed from 3.5. (Conversely, as we work on 3.6 new errors can occur - but that is the programmers life).

[QualCoder-3.5-mod.zip](https://github.com/user-attachments/files/17638470/QualCoder-3.5-mod.zip)

p.s. I think Kai and I enjoy helping and promoting QualCoder. I think as academics and being involved with research projects - finding and fixing these types of issues are a satisfying problem-solving activity.

I also appreciate that you have tried the 3.6 version, yet to be released - and found errors - as it uncovering errors helps make the software more robust.

I am curious to know how you became aware of QualCoder and - despite your initial macOS frustrations - what your colleagues and yourself think of it.

**Reply by KiwiAllan** [reply]
> I think you won't find torch with the spotlight search. Instead, run `pip list` in a command window. You will get a long list of python module installed on your system and the corresponding version number. Look for "torch" and what version it is.

HI Kai. FYI, I've just checked for torch in my system, and it doesn't appear in the list.

**Reply by kaixxx** [reply]
@Colin: Do you understand the dependencies conflict in the screenshot above? I don't get it... 
The only thing that catches my eye is that pip is trying to install very old versions of sentence-transformer. It's downloading 0.2.0, then even 0.1.0. Why? The current version is 3.2.1 and that's what I have on my system.

**Reply by KiwiAllan** [reply]
> @KiwiAllan OK Here it is - this should work fine. It is the zipped qualcoder 3.5 (I added the suffix mod to show it has been modified). This removes the speech to text option, and in doing so, removes the need for the error-prone pydub module. Regards 3.6 versus 3.5 comment. The main database (containing your coding and plain text representations of text files) is the same between the 2 versions - so interoperable. And some errors/glitches have been fixed from 3.5. (Conversely, as we work on 3.6 new errors can occur - but that is the programmers life).
> 
> [QualCoder-3.5-mod.zip](https://github.com/user-attachments/files/17638470/QualCoder-3.5-mod.zip)
> 

Yehah! 3.5 works! Thanks so much, Colin and Kai. Allan

**Reply by ccbogel** [reply]
@kaixxx  there is this:
https://stackoverflow.com/questions/71083239/unable-to-install-sentence-transformers-getting-error

I think it is trying procedurally to find a version that does not conflict, from the newest one in its list and working backwards. I presume pip was upgraded to the latest pip, prior to this lot of requirements install being done?  Because, if not, then that could be a cause of not having the latest and compatible modules ..?

**Reply by ccbogel** [reply]
@KiwiAllan Great. It was an easy fix, once I understood the cause.

**Reply by KiwiAllan** [reply]
> p.s. I think Kai and I enjoy helping and promoting QualCoder. I think as academics and being involved with research projects - finding and fixing these types of issues are a satisfying problem-solving activity.
> 
> I also appreciate that you have tried the 3.6 version, yet to be released - and found errors - as it uncovering errors helps make the software more robust.
> 
> I am curious to know how you became aware of QualCoder and - despite your initial macOS frustrations - what your colleagues and yourself think of it.

I have been, and still am even though a little long in the tooth, involved in academia (Christian theological education institutions predominantly), and appreciate collegiality expressed in all sorts of ways. Your generous, competent input over the last day has been another wonderful experience of that.

Living in Queensland, I am managing an Asia-wide qualitative research project and we need a way of undertaking the data coding and analysis. Looking around at what's available, Qualcoder came up on the list. Having a colleague who understands how it operates has led us to adopt it as our QDA platform. It will be interesting to see how things go. He's on a PC and has had no trouble installing and using it), while I'm on a Mac.

Warm regards.  Allan.

**Comment by ccbogel** [comment]
@kaixxx There is this older post: https://stackoverflow.com/questions/71083239/unable-to-install-sentence-transformers-getting-error

Another thing:
 installing the packages in a specific order. Sometimes, the order of installation can affect the resolution of dependencies. Could try installing LangChain first, then ChromaDB, and finally sentence-transformers.

**Comment by KiwiAllan** [comment]
A further query, if you don't mind. Is there a way to put a shortcut on my desktop to open Qualcoder, without going through Terminal everytime?

**Reply by ccbogel** [reply]
OK try this. As I am not that familiar with macOS.
You installed other python modules, so install this one also:
pyinstaller

Then cd to the QualCoder-3.5-mod folder

Then run this bit of code, which might take a few minutes:
pyinstaller -w -n "QualCoder-3.5-mod" --onefile qualcoder/\_\_main\_\_.py

After it has hopefully successfully finished - there should be a new folder in there called Dist

In this folder will be an executable called QualCoder-3.5-mod. 
You should be able to double-click (or mac equivalent) to make it run.

**Reply by KiwiAllan** [reply]
> pyinstaller -w -n "QualCoder-3.5-mod" --onefile qualcoder/__main__.py

Can you check this line, Colin – I'm getting 'command not found' (My QC folder is QualCoder-3.5)
![Screenshot 2024-11-06 at 10 29 30 AM](https://github.com/user-attachments/assets/71a3aff4-5078-4268-8685-cac53cb7bc68)

**Reply by ccbogel** [reply]
Try first:
pip install pyinstaller

**Reply by KiwiAllan** [reply]
'Then run this bit of code, which might take a few minutes:
pyinstaller -w -n "QualCoder-3.5-mod" --onefile qualcoder/__main__.py'

That installer worked, thanks, Colin. Then this line worked for quite a while until getting this purple/red batch of text:

![Screenshot 2024-11-06 at 10 49 16 AM](https://github.com/user-attachments/assets/0dd9dc59-f7df-4cf6-bc7e-d083c07f7e24)

**Reply by ccbogel** [reply]
ok - so it needs an image file. I though only Windows needed that.

Run again, from the same location, but like this:

pyinstaller -w -n "QualCoder-3.5-mod" --icon=QC_Logo.ico --onefile qualcoder/\_\_main\_\_.py

**Reply by KiwiAllan** [reply]
Thanks. Is it safe to remove the contents of the directory as shown in this screen shot?
![Screenshot 2024-11-06 at 11 51 34 AM](https://github.com/user-attachments/assets/42b99239-3e52-4715-bb0a-e4e717e9b1d1)

**Reply by ccbogel** [reply]
yes, because it will be replaced with the hopefully ready-to-go newer one with the image icon in it

**Reply by KiwiAllan** [reply]
> yes, because it will be replaced with the hopefully ready-to-go newer one with the image icon in it

Superb, thanks Colin. All good to go!

**Comment by kaixxx** [comment]
@KiwiAllan: Nice to hear that version 3.5 now works for you. 
It would be great, however, if you could still help us to iron out the errors in the installation process of the new version (3.6) on Mac. If you find a little bit of time, could you try to install torch with `pip install torch`, and then install the requirements again (`pip -r install requirements.txt`)?

(I will be offline for the next couple of hours, it's late over here in Europe...)

**Reply by KiwiAllan** [reply]
Happy to do so, Kai. However, I can't install torch.
![Screenshot 2024-11-06 at 10 38 21 AM](https://github.com/user-attachments/assets/407392c3-2697-4a2f-a81a-26a362fa020b)

**Reply by kaixxx** [reply]
Thank you, I think I understand now what the root of the issue is: **You are running python 3.13 which is not yet fully supported by many python libraries, torch included.** If you want to run QualCoder 3.6, you must go back to python 3.11. There is one library in particular which doesn't work with newer versions of python atm. I will check in the readme if we could make this requirement a bit clearer. 
QualCoder 3.5 doesn't use this library, so it may be fine with python 3.13.

**Reply by ccbogel** [reply]
Ahh. that makes sense

**Reply by KiwiAllan** [reply]
I'm glad I have helped to make that discovery, Kai. As I now have 3.5 working, I will skip trying to get 3.6 up and running. Cheers.

**Reply by kaixxx** [reply]
That's fine, thank you, @KiwiAllan.

**Comment by pepaganini** [comment]
Hello everyone,

It is a good surprise to have found QualCoder and to be using in my own research and also with some students who have been working with me.

After reading manuals and watched tutorials (which I appreaciate for the contributions), I wasn't able to find the solution for a problem I've been through. Everytime I try to generate a chart, it opens a new chrome tab (that's ok). But then this tab keeps running until returns that the page has not found. Also, QualCoder stops  running.

For those who needs more info to help: Windows 11 and QualCoder-3.5.exe . Other features are running well.

Cheers from Brazil.

Paulo

**Reply by ccbogel** [reply]
Hello Paulo.

Thank you for your comment.

Yes, on some computers I have had trouble with charts opening and on other computers it works perfectly. It will open charts in the default web browser (that is how the plotly library works). One suggestion is to change the default web browser to something different and see if that will work.
Apparently plotly work best in current, Chrome, Firefox or Safari web browsers. 

Potentially a really large dataset might also be a problem, but otherwise, I am not sure what else to suggest,

https://www.juliasos.com/viz/plotly-not-plotting-in-browser/#:~:text=If%20Plotly%20is%20properly%20installed%20and%20configured%2C%20but,an%20up-to-date%20version%20of%20one%20of%20these%20browsers.

regards Colin

**Comment by pepaganini** [comment]
Dear,
Thank you again for the explanation about QualCoder.
One more question that come to my mind is how can I revert when I merge two or more codes?
Thanks!

**Reply by ccbogel** [reply]
Ahh - you cannot revert. But you could open a recent backed up project. Providing you have not done a lot of extra work afterwards.
